### ISSUE TITLE
Specifies scalar values casted to match input type.

### ISSUE THREAD

**Issue by philloooo** [webmachinelearning/webnn#678]
Some of the operations have scalar parameters, e.g. [linear](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-linear), [gemm](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-gemm)'s `alpha` and `beta`.
These are passed as  `float` numbers.  I think we should specify in the processing algorithm that they get casted to match with the input type.  So that if input is float16, they also get downcast to float16.

Background is - in some cases, CoreML requires the parameter types to match. So we either downcast the scalar params to float16, or upcast input operands to float32, the latter is less ideal because on CoreML only float16 get executed on NPU.

Any concerns with adding this step in the algorithm?

**Comment by inexorabletash** [comment]
Relevant:

- #489 - defining casting generally
- #325 - and worth noting that casting fp64 &rarr; fp16 and fp64 &rarr; fp32 &rarr; fp16 may not yield the same value; that is, if an input is downcast to fp16, whether we define an input in WebIDL as `float` (fp32) or `double` (fp64) would give different results as the former implies two stage conversion from a JS Number!

**Comment by inexorabletash** [comment]
Here's one way this could be specified, using [linear()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-linear) as an example.

> 1. ...
> 1. Let _alpha_ be _options_.`alpha`, **cast** to _input's_ dataType.
> 1. Let _beta_ be _options_.`beta`, **cast** to _input's_ dataType.
> 1. ...
>> 1. ...
>> 1. Let _operator_ be an operator for the linear operation, given _alpha_ and _beta_.
>> 1. ...

And then add a definition of **cast** like:

> To **cast** a number _number_ to a given `MLOperandDataType` _dataType_, perform the following steps:
> 1. Switch on dataType
>     ↪ "float32"
>       Return the result of [converting a JavaScript value to an IDL unrestricted float value](https://webidl.spec.whatwg.org/#js-unrestricted-float) with _number_.
>     ↪ "float16"
>       TODO
>     ↪ "int32"
>       Return the result of [converting a JavaScript value to an IDL long value](https://webidl.spec.whatwg.org/#js-long) with _number_.
>     ↪ "uint32"
>       Return the result of [converting a JavaScript value to an IDL unsigned long value](https://webidl.spec.whatwg.org/#js-unsigned-long) with _number_.
>     ↪ "int64"
>       Return the result of [converting a JavaScript value to an IDL long long value](https://webidl.spec.whatwg.org/#js-long-long) with _number_.
>     ↪ "uint64"
>       Return the result of [converting a JavaScript value to an IDL unsigned long long value](https://webidl.spec.whatwg.org/#js-unsigned-long-long) with _number_.
>     ↪ "int8"
>       Return the result of [converting a JavaScript value to an IDL byte value](https://webidl.spec.whatwg.org/#js-byte) with _number_.
>     ↪ "uint8"
>       Return the result of [converting a JavaScript value to an IDL octet](https://webidl.spec.whatwg.org/#js-octet) with _number_.

With caveats:
* The input to this only makes sense if it's a (unrestricted) double which is equivalent to the JavaScript number type. See #647 for discussions around BigInt support, which is necessary for full-precision int64 inputs.
* IDL doesn't have a half/float16 type (yet?) - we can copy/paste/tweak, reference https://tc39.es/proposal-float16array, etc
* This intentionally gives the conversion the same result as if we had explicit IDL for the types, for example:

```webidl
dictionary MLLinearFloat32Options {
  float alpha = 1;
  float beta = 1;
};
dictionary MLLinearFloat16Options {
  half alpha = 1; // this type doesn't actually exist
  half beta = 1; // this type doesn't actually exist
};
partial interface MLGraphBuilder {
  MLOperand linearFloat32(MLOperand input, optional MLLinearFloat32Options options = {});
  MLOperand linearFloat16(MLOperand input, optional MLLinearFloat16Options options = {});
};
```
... which is probably more relevant for integer types, because IDL+JS have well defined if sometimes surprising behavior here. See #489 for discussion about casting within _backends_, but at least at the IDL/interface level we should behave predictably.

**Comment by inexorabletash** [comment]
More notes:

- I have a more detailed local sketch of the cast algorithm that doesn't rely on reaching into non-exported parts of WebIDL, and addresses some of the limitations (bigint, float16, etc). We may want to roll this into #647
- Is this a comprehensive list of where we'd want to directly invoke "cast" or are there more? Are there instances here we don't need to cast? (resample2d seems sketchy to me given #474)
  - [constant(value, type)](https://webmachinelearning.github.io/webnn/#dom-mlgraphbuilder-constant-value-type) - [value](https://webmachinelearning.github.io/webnn/#dom-mlgraphbuilder-constant-value-type-value)
  - [batchNormalization()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-batchnorm) - [epsilon](https://webmachinelearning.github.io/webnn/#dom-mlbatchnormalizationoptions-epsilon)
  - [clamp()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-clamp) (both op and activation) - [minValue](https://webmachinelearning.github.io/webnn/#dom-mlclampoptions-minvalue) and [maxValue](https://webmachinelearning.github.io/webnn/#dom-mlclampoptions-maxvalue)
  - [elu()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-elu) (both op and activation) - [alpha](https://webmachinelearning.github.io/webnn/#dom-mleluoptions-alpha)
  - [gemm()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-gemm) - [alpha](https://webmachinelearning.github.io/webnn/#dom-mlgemmoptions-alpha) and [beta](https://webmachinelearning.github.io/webnn/#dom-mlgemmoptions-beta)
  - [hardSigmoid()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-hard-sigmoid) (both op and activation) - [alpha](https://webmachinelearning.github.io/webnn/#dom-mlhardsigmoidoptions-alpha) and [beta](https://webmachinelearning.github.io/webnn/#dom-mlhardsigmoidoptions-beta)
  - [instanceNormalization()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-instancenorm) - [epsilon](https://webmachinelearning.github.io/webnn/#dom-mlinstancenormalizationoptions-epsilon)
  - [layerNormalization()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-layernorm) - [epsilon](https://webmachinelearning.github.io/webnn/#dom-mllayernormalizationoptions-epsilon)
  - [leakyRelu()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-leakyrelu) (both op and activation) - [alpha](https://webmachinelearning.github.io/webnn/#dom-mlleakyreluoptions-alpha)
  - [linear()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-linear) (both op and activation) - [alpha](https://webmachinelearning.github.io/webnn/#dom-mllinearoptions-alpha) and [beta](https://webmachinelearning.github.io/webnn/#dom-mllinearoptions-beta)
  - [pad()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-pad) - [value](https://webmachinelearning.github.io/webnn/#dom-mlpadoptions-value)
  - [resample2d()](https://webmachinelearning.github.io/webnn/#api-mlgraphbuilder-resample2d-method) - [scales](https://webmachinelearning.github.io/webnn/#dom-mlresample2doptions-scales)

**Comment by inexorabletash** [comment]
Also, what conversion options would we use for these? Specifically:
  * For casts to a floating point type, we can make this restricted (disallow +Infinity/-Infinity/NaN) or allow them
  * For casts to an integral type, we can make them act like [EnforceRange] (e.g. -1 cast to uint8 throws), [Clamp] (e.g. -1 cast to uint8 yields 0), or neither (C-style modulus e.g. -1 cast to uint8 yields 255)

This sounds like #489 but that issue is specifically about the casting operator, which might depend on the underlying platform. This issue is talking about the JS interface which is going to be implemented by the user agent.

**Comment by inexorabletash** [comment]
And then a further note for consideration: if we're being explicit about casting everywhere, are there any places that are currently `float` (that we're considering making `double`, see #325) that could be `MLNumber` as well so a developer could pass either 123456789123456789 or 123456789123456789n ?

This is fairly moot as all of the relevant ops accept only floating point inputs, so whatever the developer supplies will be ultimately cast to a float32 or float16, and so specifying the input as `double` rather than `MLNumber` doesn't lose detail, except the an exception will be thrown if 123456789123456789n is passed.

**Comment by inexorabletash** [comment]
Unsorted notes from internal discussion w/ @a-sully and @philloooo:

For batchNorm's epsilon. 
- WebNN specfies that the input may be float16 or float32
- in [DML](https://learn.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_batch_normalization_operator_desc), epsilon is always a float32
- in [CoreML](https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS17.normalization.batch_norm), epsilon can be float16 or float32
- TFLite doesn't naively support batchNorm so it would be polyfilled (and then we'd need to reason about the data types in more detail)

And:
- DML is basically float32 everywhere (elu's alpha, gemm's alpha/beta, hardSigmoid, etc); if the alpha is float16 but the input's data type is float16 what ends up happening?
- CoreML in iOS15 it required these parameters to be the same dtype as the input; only in iOS17 was this restriction relaxed for many operators

So overall, the _"cast parameters to the same data type as the input tensors"_ approach may be okay and likely the least surprising for developers even if the result gets upcast (e.g. fp16 to fp32) again, but there's a lot of nuance to investigate.
